# Task
Consigne 1 : Modélisation en ML

Here is all the data you need:
"submission.csv"

## Data loading

### Subtask:
Load the data from "submission.csv" into a pandas DataFrame.


**Reasoning**:
Load the data from "submission.csv" into a pandas DataFrame and display its first few rows and shape.



In [1]:
import pandas as pd

df = pd.read_csv('submission.csv')
display(df.head())
print(df.shape)

,textID,selected_text
0,f87dea47db,Last session of the day http://twitpic.com/67ezh
1,96d74cb729,exciting
2,eee518ae67,Recession
3,01082688c6,happy bday!
4,33987a8ee5,I like it!!


(3534, 2)


## Data exploration

### Subtask:
Explore the loaded data to understand its structure, features, and potential missing values. Determine the target variable and relevant features.


**Reasoning**:
Explore the data by examining data types, missing values, descriptive statistics, unique values, and text analysis, then identify the target variable and relevant features.



In [2]:
# Examine data types
print(df.dtypes)

# Check for missing values
print(df.isnull().sum())
print(df.isnull().sum() / len(df) * 100)

# Descriptive statistics for numerical columns (if any)
print(df.describe(include='all'))

# Analyze unique values in each column
for col in df.columns:
    print(f"Unique values in '{col}': {df[col].unique()}")
    print(f"Frequency of unique values in '{col}': \n{df[col].value_counts()}")

# Basic text analysis (for 'selected_text' column)
if df['selected_text'].dtype == 'object':
  df['selected_text_length'] = df['selected_text'].astype(str).apply(len)
  print(f"Average length of selected text: {df['selected_text_length'].mean()}")
  #Further analysis can be added if needed.  Example: Unique words, special characters etc.

# Identify target variable and relevant features
# The target variable is likely 'selected_text' as it seems to represent the outcome or result of a task.
# 'textID' is likely a unique identifier and could be used as an index.

# Summarize findings
print("\nSummary:")
print("The dataset contains two columns: 'textID' (unique identifier) and 'selected_text' (likely the target variable).")
print("There are no missing values. The descriptive statistics for the 'selected_text' column provides the information on the lengths of the texts, among other things.")

textID           object
selected_text    object
dtype: object
textID           0
selected_text    0
dtype: int64
textID           0.0
selected_text    0.0
dtype: float64
            textID selected_text
count         3534          3534
unique        3534          2977
top     469e15c5a8          good
freq             1            50
Unique values in 'textID': ['f87dea47db' '96d74cb729' 'eee518ae67' ... '6332da480c' 'df1baec676'
 '469e15c5a8']
Frequency of unique values in 'textID': 
textID
469e15c5a8    1
f87dea47db    1
96d74cb729    1
eee518ae67    1
01082688c6    1
             ..
33f19050cf    1
00d5195223    1
9fce30159a    1
ae93ad52a0    1
504e45d9d9    1
Name: count, Length: 3534, dtype: int64
Unique values in 'selected_text': ['Last session of the day  http://twitpic.com/67ezh' 'exciting'
 'Recession' ...
 'Friday evening......what to do, what to do.  I have to idea.'
 ' I know what you mean. My little dog is sinking into depression... he wants to move someplace tropical'
 'I 

## Data preparation

### Subtask:
Prepare the data for modeling by handling potential issues like inconsistent data types and further exploring the 'selected_text' column.


**Reasoning**:
Convert the `textID` column to a categorical data type, as it represents unique identifiers and is unlikely to have any ordinal relationship.  Further investigate the `selected_text` column for preprocessing needs.



In [3]:
# Convert textID to categorical
df['textID'] = df['textID'].astype('category')

# Lowercase the selected text
df['selected_text_lower'] = df['selected_text'].str.lower()

# Remove punctuation
import string
df['selected_text_no_punctuation'] = df['selected_text_lower'].apply(lambda x: x.translate(str.maketrans('', '', string.punctuation)))

#Further preprocessing steps can be added here, like removing stopwords, stemming/lemmatization, etc.
display(df.head())

,textID,selected_text,selected_text_length,selected_text_lower,selected_text_no_punctuation
0,f87dea47db,Last session of the day http://twitpic.com/67ezh,49,last session of the day http://twitpic.com/67ezh,last session of the day httptwitpiccom67ezh
1,96d74cb729,exciting,8,exciting,exciting
2,eee518ae67,Recession,9,recession,recession
3,01082688c6,happy bday!,12,happy bday!,happy bday
4,33987a8ee5,I like it!!,11,i like it!!,i like it


## Feature engineering

### Subtask:
Engineer new features from the preprocessed text data.


**Reasoning**:
Calculate the length, word count, character count, and average word length of the 'selected_text_no_punctuation' column and store them in new columns.



In [4]:
df['text_length'] = df['selected_text_no_punctuation'].str.len()
df['word_count'] = df['selected_text_no_punctuation'].str.split().str.len()
df['char_count'] = df['selected_text_no_punctuation'].str.len()
df['avg_word_length'] = df['char_count'] / df['word_count']
display(df.head())

,textID,selected_text,selected_text_length,selected_text_lower,selected_text_no_punctuation,text_length,word_count,char_count,avg_word_length
0,f87dea47db,Last session of the day http://twitpic.com/67ezh,49,last session of the day http://twitpic.com/67ezh,last session of the day httptwitpiccom67ezh,44,6,44,7.333333
1,96d74cb729,exciting,8,exciting,exciting,8,1,8,8.000000
2,eee518ae67,Recession,9,recession,recession,9,1,9,9.000000
3,01082688c6,happy bday!,12,happy bday!,happy bday,11,2,11,5.500000
4,33987a8ee5,I like it!!,11,i like it!!,i like it,9,3,9,3.000000


## Data splitting

### Subtask:
Split the data into training and testing sets.


**Reasoning**:
Split the data into training and testing sets using `train_test_split`. Define features (X) and target variable (y) as instructed.  Use `text_length`, `word_count`, `char_count`, and `avg_word_length` as features. Use `selected_text` as target.  Remove `textID` and other unnecessary columns from the feature set. Set the random state for reproducibility.



In [5]:
from sklearn.model_selection import train_test_split

# Define features (X) and target variable (y)
X = df[['text_length', 'word_count', 'char_count', 'avg_word_length']]
y = df['selected_text']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Display shapes of the resulting sets to verify the split
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (2827, 4)
X_test shape: (707, 4)
y_train shape: (2827,)
y_test shape: (707,)


## Model training

### Subtask:
Train a RandomForestRegressor model on the training data.


**Reasoning**:
Train a RandomForestRegressor model on the training data.  Since the target variable `y_train` is text, we need to convert it to numerical representations before fitting the model.  I will use Label Encoding for this purpose.



In [6]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.preprocessing import LabelEncoder

# Initialize LabelEncoder
le = LabelEncoder()

# Fit and transform the target variable
y_train_encoded = le.fit_transform(y_train)

# Initialize the RandomForestRegressor model
model = RandomForestRegressor(n_estimators=100, random_state=42)

# Train the model
model.fit(X_train, y_train_encoded)

RandomForestRegressor(random_state=42)

## Model optimization

### Subtask:
Optimize the RandomForestRegressor model.


**Reasoning**:
Import GridSearchCV and optimize the RandomForestRegressor model using GridSearchCV with specified parameters and cross-validation.



In [7]:
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestRegressor
from sklearn.preprocessing import LabelEncoder

# Initialize LabelEncoder
le = LabelEncoder()

# Fit and transform the target variable
y_train_encoded = le.fit_transform(y_train)

# Define the parameter grid
param_grid = {
    'n_estimators': [50, 100, 200],
    'max_depth': [None, 10, 20],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

# Initialize the RandomForestRegressor model
model = RandomForestRegressor(random_state=42)

# Initialize GridSearchCV
grid_search = GridSearchCV(estimator=model, param_grid=param_grid, scoring='neg_mean_squared_error', cv=5)

# Fit GridSearchCV to the training data
grid_search.fit(X_train, y_train_encoded)

# Get the best estimator
best_model = grid_search.best_estimator_

# Print the best parameters and best score
print("Best parameters:", grid_search.best_params_)
print("Best score:", grid_search.best_score_)

Best parameters: {'max_depth': 10, 'min_samples_leaf': 4, 'min_samples_split': 10, 'n_estimators': 200}
Best score: -385613.6263178505


## Model evaluation

### Subtask:
Evaluate the performance of the best model on the test set.


**Reasoning**:
Evaluate the best model's performance on the test set by predicting on X_test, inverse transforming the predictions and y_test using the label encoder, and calculating MSE and R2 score.



In [8]:
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import LabelEncoder

# Initialize LabelEncoder
le = LabelEncoder()

# Fit and transform y_test
y_test_encoded = le.fit_transform(y_test)

# Predict on the test set
y_pred_encoded = best_model.predict(X_test)

# Inverse transform the predictions and y_test
y_pred = le.inverse_transform(y_pred_encoded.astype(int))
y_test_original = le.inverse_transform(y_test_encoded)

# Calculate MSE
mse = mean_squared_error(y_test_encoded, y_pred_encoded)

# Calculate R2 score
r2 = r2_score(y_test_encoded, y_pred_encoded)


# Print the MSE and R2 values
print(f"Mean Squared Error (MSE): {mse}")
print(f"R-squared (R2) Score: {r2}")

ValueError: y contains previously unseen labels: [ 657  660  690  694  699  701  702  712  714  719  720  721  733  737
  740  742  754  769  770  771  772  777  780  781  798  803  804  806
  807  815  816  819  821  823  825  827  828  830  841  843  849  854
  855  857  858  860  861  863  865  866  867  868  873  877  878  880
  884  887  889  890  891  895  897  898  899  900  901  902  905  907
  908  913  914  915  919  920  921  922  926  928  930  931  932  942
  943  946  950  951  954  955  956  957  958  959  962  966  971  975
  976  980  984  985  986  987  988  989  994  996  997  999 1000 1002
 1003 1004 1009 1010 1011 1015 1018 1019 1020 1022 1025 1026 1029 1034
 1038 1040 1042 1043 1044 1045 1047 1048 1049 1053 1054 1056 1061 1063
 1067 1069 1070 1071 1074 1075 1076 1082 1083 1087 1091 1093 1094 1095
 1096 1097 1099 1100 1101 1112 1115 1118 1119 1122 1124 1125 1126 1128
 1131 1132 1134 1135 1136 1139 1142 1144 1146 1148 1151 1152 1158 1160
 1164 1166 1167 1173 1174 1176 1177 1181 1182 1185 1186 1188 1189 1191
 1194 1205 1209 1211 1212 1213 1218 1220 1227 1229 1231 1239 1242 1244
 1245 1246 1250 1251 1254 1255 1258 1259 1260 1266 1272 1276 1280 1284
 1289 1300 1309 1314 1320 1321 1326 1333 1337 1346 1352 1355 1356 1362
 1374 1376 1378 1381 1391 1401 1410 1427 1439 1469 1501 1510 1524 1527
 1534 1540 1575 1577 1579 1582 1588 1600 1612 1625 1640 1645 1653 1663
 1673 1681 1686 1700 1742 1772 1775 1783 1936]

**Reasoning**:
The previous code failed due to unseen labels in `y_test` during inverse transformation.  The label encoder fitted on `y_train` does not contain all the labels present in `y_test`. We need to fit the label encoder on the combined set of labels from `y_train` and `y_test`. Then, transform both `y_train` and `y_test` using this fitted label encoder.  This will ensure that both the training and test data use the same set of labels for encoding.  Then, proceed with model evaluation.



In [9]:
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

# Combine labels from y_train and y_test
all_labels = np.concatenate((y_train, y_test))

# Initialize LabelEncoder
le = LabelEncoder()

# Fit LabelEncoder on all labels
le.fit(all_labels)

# Transform y_train and y_test using the fitted LabelEncoder
y_train_encoded = le.transform(y_train)
y_test_encoded = le.transform(y_test)

# Predict on the test set
y_pred_encoded = best_model.predict(X_test)

# Calculate MSE
mse = mean_squared_error(y_test_encoded, y_pred_encoded)

# Calculate R2 score
r2 = r2_score(y_test_encoded, y_pred_encoded)

# Print the MSE and R2 values
print(f"Mean Squared Error (MSE): {mse}")
print(f"R-squared (R2) Score: {r2}")

Mean Squared Error (MSE): 665314.2418291754
R-squared (R2) Score: 0.08676425882965821


## Summary:

### 1. Q&A

No explicit questions were provided in the data analysis task.  However, the steps implicitly ask:

*   What is the target variable?  Answer: `selected_text`.
*   How can we prepare the data for modeling? Answer: Convert `textID` to categorical, lowercase the `selected_text`, remove punctuation, and potentially further preprocess the text (e.g., remove stop words, stem/lemmatize).
*   What features can be engineered from the text? Answer:  `text_length`, `word_count`, `char_count`, and `avg_word_length`.
*   How does the model perform on the test data? Answer: The model achieves an MSE of 665314.24 and an R2 score of 0.0868.
*   What are the best hyperparameters for the model? Answer: `{'max_depth': 10, 'min_samples_leaf': 4, 'min_samples_split': 10, 'n_estimators': 200}`.

### 2. Data Analysis Key Findings

*   **No Missing Values:**  The dataset contains no missing values in either the `textID` or `selected_text` columns.
*   **Target Variable:** `selected_text` is identified as the target variable.
*   **Average Text Length:** The average length of the text in the `selected_text` column is approximately 42.12 characters.
*   **Model Optimization:**  Grid search identified optimal hyperparameters for the `RandomForestRegressor`, resulting in a best score (negative mean squared error) of -385613.63 on the training set.  The best parameters were `{'max_depth': 10, 'min_samples_leaf': 4, 'min_samples_split': 10, 'n_estimators': 200}`.
*   **Model Evaluation:** The optimized model achieved an MSE of 665314.24 and an R2 score of 0.0868 on the test set.

### 3. Insights or Next Steps

*   **Feature Engineering:** Explore more sophisticated text features (e.g., TF-IDF, word embeddings) to potentially improve model performance. The current features are very basic and may not capture the nuances of the text data.  The low R2 score suggests that the model is not explaining much of the variance in the target variable.
*   **Model Selection:**  Consider alternative models better suited for text data, such as recurrent neural networks (RNNs) or transformers.  The low R2 score suggests that the current model might not be appropriate for this task.  Given the nature of the task, which seems to involve selecting text, sequence models might be more appropriate.
